snuba/web/rpc/common.py

```diff
@@ -0,0 +1,281 @@
+
+        if exp.function_name == "and":
+            return combine_and_conditions(exp.parameters)
+        elif exp.function_name == "or":
+            return combine_or_conditions(exp.parameters)
+        else:
+            return exp
+
+    query.transform_expressions(transform)
+
+
+# These are the columns which aren't stored in attr_str_ nor attr_num_ in clickhouse
+NORMALIZED_COLUMNS: Final[Mapping[str, AttributeKey.Type.ValueType]] = {
+    "project_name": AttributeKey.Type.TYPE_STRING,  # extra special column, derived from a context sent with the request
+    "organization_id": AttributeKey.Type.TYPE_INT,
+    "project_id": AttributeKey.Type.TYPE_INT,
+    "service": AttributeKey.Type.TYPE_STRING,
+    "span_id": AttributeKey.Type.TYPE_INT,
+    "parent_span_id": AttributeKey.Type.TYPE_INT,
+    "segment_id": AttributeKey.Type.TYPE_INT,
+    "segment_name": AttributeKey.Type.TYPE_STRING,
+    "is_segment": AttributeKey.Type.TYPE_BOOLEAN,
+    "duration_ms": AttributeKey.Type.TYPE_INT,
+    "exclusive_time_ms": AttributeKey.Type.TYPE_INT,
+    "retention_days": AttributeKey.Type.TYPE_INT,
+    "name": AttributeKey.Type.TYPE_STRING,
+    "sample_weight": AttributeKey.Type.TYPE_FLOAT,
+    "timestamp": AttributeKey.Type.TYPE_UNSPECIFIED,
+    "start_timestamp": AttributeKey.Type.TYPE_UNSPECIFIED,
+    "end_timestamp": AttributeKey.Type.TYPE_UNSPECIFIED,
+}
+
+# Columns stored as integers that are usually presented to users as hex strings
+HEX_ID_COLUMNS: Final[Set[str]] = {"span_id", "parent_span_id", "segment_id"}
+
+TIMESTAMP_COLUMNS: Final[Set[str]] = {"timestamp", "start_timestamp", "end_timestamp"}
+
+
+def attribute_key_to_expression(
+    attr_key: AttributeKey, context: Optional[AttributeKeyTransformContext]
+) -> Expression:
+    if attr_key.type == AttributeKey.Type.TYPE_UNSPECIFIED:
+        raise BadSnubaRPCRequestException(
+            f"attribute key {attr_key.name} must have a type specified"
+        )
+    # Snuba doesn't have access to postgres, which stores project_id to project_name mappings.
+    # This context object lets the caller specify that mapping, so that they can (for example) order by project name
+    if (
+        attr_key.name == "project_name"
+        and attr_key.type == AttributeKey.Type.TYPE_STRING
+    ):
+        if context is None or len(context.project_ids_to_names) == 0:
+            raise BadSnubaRPCRequestException(
+                "If you request project_name, the AttributeKeyTransformContext must be "
+                "sent. Snuba doesn't have access to project names directly."
+            )
+        return f.transform(
+            column("project_id"),
+            literals_array(
+                None, [literal(k) for k in context.project_ids_to_names.keys()]
+            ),
+            literals_array(
+                None, [literal(v) for v in context.project_ids_to_names.values()]
+            ),
+            literal("unknown"),
+            alias="project_name",
+        )
```

Comment:
Why are we doing this? What's the problem with querying by project id?

Reply:
people want to do things like order by, which wouldn't make sense if it was ordering by project id

Reply:
People are trying to order by project name specifically?

Reply:
that's what product said

Reply:
there's a table that can contain project name, and if they click the carat they would like it to actually order by project name - this logic currently lives in the sentry backend, we can probably optimize it by making a `Join` table for project_name <> project_id, since that seems like it'll come up again in the future